src/main/java/seedu/address/logic/commands/CraftItemCommand.java

```diff
@@ -0,0 +1,206 @@
+        this.hasCraftedExcess = false;
+        this.hasDefaultIndex = true;
+    }
+
+    @Override
+    public CommandResult execute(Model model) throws CommandException {
+        requireNonNull(model);
+
+        if (productQuantity == 0) {
+            throw new CommandException(MESSAGE_INVALID_PRODUCT_QUANTITY);
+        }
+
+        // fetch the item
+        List<Item> tempItemList = new ArrayList<>(model.getFilteredItemList());
+        // filter to only get matching items
+        tempItemList.removeIf(x -> !x.getName().equals(itemName));
+        if (tempItemList.isEmpty()) {
+            throw new CommandException(MESSAGE_ITEM_NOT_FOUND);
+        }
+        Item itemToCraft;
+        itemToCraft = tempItemList.stream()
+                .findFirst() // Get the first (and only) item matching or else throw Error
+                .orElseThrow(() -> new CommandException(MESSAGE_ITEM_NOT_FOUND));
+
+        // fetch the recipe
+        List<Recipe> tempRecipeList = new ArrayList<>(model.getFilteredRecipeList());
+        // filter to only get matching recipes
+        tempRecipeList.removeIf(x -> !x.getProductName().equals(itemName));
+        if (tempRecipeList.isEmpty()) {
+            throw new CommandException(MESSAGE_RECIPE_NOT_FOUND);
+        }
+        Recipe recipeToUse;
+        try {
+            recipeToUse = tempRecipeList.get(index.getZeroBased());
+        } catch (IndexOutOfBoundsException e) {
+            throw new CommandException(MESSAGE_INDEX_OUT_OF_RANGE);
+        }
+
+        int recipeProductQuantity = recipeToUse.getProductQuantity().number;
+        // check if quantity to craft is not an exact multiple of recipe product quantity
+        int intendedQuantity = productQuantity;
+        if (productQuantity % recipeProductQuantity != 0) {
+            // round up to craft extra
+            productQuantity = ((productQuantity / recipeProductQuantity) + 1) * recipeProductQuantity;
+            hasCraftedExcess = true;
+        }
+
+        IngredientList ingredientList = recipeToUse.getIngredients();
+        ArrayList<Item> itemList = new ArrayList<>(model.getFilteredItemList());
+        // Storage into hashmap to use for updating
+        HashMap<Integer, Integer> requiredIngredients = new HashMap<>();
+
+        // check if there is enough of each ingredient in inventory
+        if (!checkIngredients(ingredientList, itemList, requiredIngredients, recipeProductQuantity)) {
+            throw new CommandException(MESSAGE_INSUFFICIENT_INGREDIENTS);
+        }
+
+        // update ingredients, decrease each by quantity required since ingredients are consumed
+        requiredIngredients.forEach((itemId, quantityRequired) -> {
+            try {
+                new AddQuantityToItemCommand(itemList.get(itemId).getName(), -quantityRequired).execute(model);
+            } catch (CommandException e) {
+                // change of quantity should never fail
+                assert false;
```

Comment:
change to more informative error message, at least the itemId of the item that caused the error (even if it never hits) - but keep documentation